lib/dl_api_lib/dl_api_lib_tests/db/data_api/test_join_optimization.py

```diff
@@ -0,0 +1,140 @@
+import pytest
+
+from dl_api_client.dsmaker.primitives import Dataset
+from dl_api_lib_tests.db.base import DefaultApiTestBase
+
+
+class TestDatasetJoinOptimization(DefaultApiTestBase):
+    @pytest.fixture(scope="function")
+    def dataset_for_join_optimization_tests(self, control_api, saved_connection_id, dataset_params):
+        ds = Dataset()
+
+        # Create 3 identical sources & corresponding avatars
+        for source_idx in (1, 2, 3):
+            ds.sources[f"source_{source_idx}"] = ds.source(
+                connection_id=saved_connection_id,
+                **dataset_params,
+            )
+            ds.source_avatars[f"avatar_{source_idx}"] = ds.sources[f"source_{source_idx}"].avatar()
+
+        # Create relations
+        ds.avatar_relations["relation_1"] = (
+            ds.source_avatars["avatar_1"]
+            .join(ds.source_avatars["avatar_2"])
+            .on(
+                ds.col("customer_id") == ds.col("customer_id"),
+            )
+        )
+        ds.avatar_relations["relation_2"] = (
+            ds.source_avatars["avatar_1"]
+            .join(ds.source_avatars["avatar_3"])
+            .on(
+                ds.col("customer_id") == ds.col("customer_id"),
+            )
+        )
+
+        ds = control_api.apply_updates(dataset=ds).dataset
+        ds = control_api.save_dataset(dataset=ds).dataset
+
+        # Resulting avatar tree:
+        # avatar_1 ── avatar_2
+        # │
+        # └── avatar_3
+        # Fields associated with avatar_2 have a suffix " (1)", same for avatar_3, no suffix for avatar_1
+
+        yield ds
+
+        control_api.delete_dataset(ds.id)
+
+    def test_join_optimization_no_required(self, data_api, dataset_for_join_optimization_tests):
+        """Selecting from a dataset without any required relations and making sure there are no JOINs"""
+
+        ds = dataset_for_join_optimization_tests
+
+        result_resp = data_api.get_result(
+            dataset=ds,
+            fields=[
+                ds.find_field(title="customer_id (1)"),  # Requesting a field associated with avatar_2
+            ],
+        )
+        assert result_resp.status_code == 200, result_resp.json
+        sent_query: str = result_resp.json["blocks"][0]["query"]
+        joins_count = sent_query.count("INNER JOIN")
```

Comment:
maybe count `JOIN`s (without `INNER`), just in case there will be test using other types?